internal/services/grpc/middleware/rate_limit.go

```diff
@@ -0,0 +1,150 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type HatchetApiTokenRateLimiter struct {
+	eventsLimiter     *rate.Limiter
+	dispatcherLimiter *rate.Limiter
+	workflowLimiter   *rate.Limiter
+}
+
+type HatchetRateLimiter struct {
+	mu           sync.Mutex
+	rateLimiters map[string]*HatchetApiTokenRateLimiter
+	rate         rate.Limit
+	burst        int
+	l            *zerolog.Logger
+}
+
+func (rl *HatchetRateLimiter) GetOrCreateTenantRateLimiter(rateLimitToken string) *HatchetApiTokenRateLimiter {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+
+	if _, ok := rl.rateLimiters[rateLimitToken]; !ok {
+		rl.rateLimiters[rateLimitToken] = &HatchetApiTokenRateLimiter{
+			eventsLimiter:     rate.NewLimiter(rl.rate, rl.burst),
+			dispatcherLimiter: rate.NewLimiter(rl.rate, rl.burst),
+			workflowLimiter:   rate.NewLimiter(rl.rate, rl.burst),
+		}
+	}
+
+	return rl.rateLimiters[rateLimitToken]
+}
+
+func NewHatchetRateLimiter(r rate.Limit, b int, l *zerolog.Logger) *HatchetRateLimiter {
+	return &HatchetRateLimiter{
+		rateLimiters: make(map[string]*HatchetApiTokenRateLimiter),
+		rate:         r,
+		burst:        b,
+		l:            l,
+	}
+}
+
+// Limit is called before each request is processed. It should return an error if rate-limited.
+func (r *HatchetRateLimiter) Limit(ctx context.Context) error {
+
+	serviceName, ok := ctx.Value(grpcServiceName).(string)
+	if !ok {
+		return status.Errorf(codes.Internal, "no server in context")
+	}
+
+	fmt.Println("the service is", serviceName)
+
+	rateLimitToken := ctx.Value("rate_limit_token").(string)
+
+	if rateLimitToken == "" {
+		return status.Errorf(codes.Unauthenticated, "no rate limit token found")
+	}
+
+	switch matchServiceName(serviceName) {
+	case "dispatcher":
```

Comment:
I don't think we should be rate-limiting the dispatcher, or at least the default rate limit should be higher for the dispatcher. We are not currently seeing issues on this service.

Reply:
ok, set it to 10X the provided rate for the dispatcher - probably best to have some rate limit and a 10X ratio seems safe enough ?

Reply:
sounds good, thanks!